lib/core/stores/local.ts

```diff
@@ -0,0 +1,364 @@
+import {
+  DurablePromise,
+  PendingPromise,
+  ResolvedPromise,
+  RejectedPromise,
+  CanceledPromise,
+  TimedoutPromise,
+  isPendingPromise,
+  isResolvedPromise,
+  isRejectedPromise,
+  isCanceledPromise,
+} from "../promise";
+import { IPromiseStore } from "../store";
+import { IEncoder } from "../encoder";
+import { Base64Encoder } from "../encoders/base64";
+import { ErrorCodes, ResonateError } from "../error";
+
+export class LocalPromiseStore implements IPromiseStore {
+  private readonly storeName = "promises";
+  private db: IDBDatabase;
+
+  private constructor(
+    db: IDBDatabase,
+    private encoder: IEncoder<string, string> = new Base64Encoder(),
+  ) {
+    this.db = db;
+  }
+
+  static async getInstance(encoder?: IEncoder<string, string>): Promise<LocalPromiseStore> {
+    return new Promise<LocalPromiseStore>((resolve, reject) => {
+      const request = indexedDB.open("resonateDB", 1);
+
+      request.onerror = () => {
+        reject(request.error);
+      };
+
+      request.onsuccess = () => {
+        const db = request.result;
+        resolve(new LocalPromiseStore(db, encoder));
+      };
+
+      request.onupgradeneeded = (event) => {
```

Comment:
If this callback is invoked, will `onsuccess` be called afterwards? Just want to make sure the promise is eventually resolved/rejected.

Reply:
Yes, if the onupgradeneeded callback is invoked, the onsuccess callback will be called afterward.